File: src/main/java/org/tobi29/scapes/engine/gui/GuiComponentEditableText.java
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.gui;

import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.graphics.*;
import org.tobi29.scapes.engine.utils.Pair;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.math.FastMath;
import org.tobi29.scapes.engine.utils.math.vector.Vector2;

import java.util.List;

public class GuiComponentEditableText extends GuiComponentHeavy {
    protected final GuiController.TextFieldData data =
            new GuiController.TextFieldData();
    protected final int maxLength;
    protected final float r, g, b, a;
    protected boolean active, focused;
    protected List<Pair<Model, Texture>> vaoCursor, vaoSelection;
    protected GuiComponentText.TextFilter textFilter = str -> str;

    public GuiComponentEditableText(GuiLayoutData parent, String text) {
        this(parent, text, 1.0f, 1.0f, 1.0f, 1.0f);
    }

    public GuiComponentEditableText(GuiLayoutData parent, String text, float r,
            float g, float b, float a) {
        this(parent, text, Integer.MAX_VALUE, r, g, b, a);
    }

    public GuiComponentEditableText(GuiLayoutData parent, String text,
            int maxLength) {
        this(parent, text, maxLength, 1.0f, 1.0f, 1.0f, 1.0f);
    }

    public GuiComponentEditableText(GuiLayoutData parent, String text,
            int maxLength, float r, float g, float b, float a) {
        super(parent);
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
        data.text.append(text);
        this.maxLength = maxLength;
        data.cursor = data.text.length();
        dirty();
    }

    public GuiController.TextFieldData data() {
        return data;
    }

    public boolean active() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    public String text() {
        return data.text.toString();
    }

    public void setText(String text) {
        if (!data.text.toString().equals(text)) {
            data.text.setLength(0);
            data.text.append(text);
            dirty();
        }
    }

    public void setTextFilter(GuiComponentText.TextFilter textFilter) {
        this.textFilter = textFilter;
        dirty();
    }

    @Override
    protected void updateMesh(GuiRenderer renderer, Vector2 size) {
        if (data == null) {
            return;
        }
        FontRenderer font = gui.style().font();
        String text = data.text.toString();
        font.render(FontRenderer.to(renderer, r, g, b, a),
                textFilter.filter(text), size.floatY(), size.floatX());
        int cursor = data.cursor;
        int selectionStart = data.selectionStart;
        int selectionEnd = data.selectionEnd;
        cursor = FastMath.clamp(cursor, 0, text.length());
        selectionStart = FastMath.clamp(selectionStart, 0, text.length());
        selectionEnd = FastMath.clamp(selectionEnd, 0, text.length());
        GuiRenderBatch batch = new GuiRenderBatch(renderer.pixelSize());
        font.render(FontRenderer.to(batch, 0.0f - size.floatY() * 0.1f,
                0.0f - size.floatY() * 0.1f, 1.0f, 1.0f, 1.0f, 1.0f),
                text.substring(0, cursor) + '|', size.floatY(),
                size.floatY() * 1.2f, size.floatY(), Float.MAX_VALUE, cursor,
                cursor + 1);
        vaoCursor = batch.finish();
        font.render(FontRenderer
                        .to(batch, 0.0f, 0.0f, true, 1.0f, 1.0f, 1.0f, 1.0f), text,
                size.floatY(), size.floatY(), size.floatY(), size.floatX(),
                selectionStart, selectionEnd);
        vaoSelection = batch.finish();
    }

    @Override
    protected void updateComponent(ScapesEngine engine, double delta) {
        if (active) {
            if (!focused) {
                engine.guiController().focusTextField(data, false);
                focused = true;
            }
            size(engine).ifPresent(size -> {
                if (engine.guiController().processTextField(data, false)) {
                    if (data.text.length() > maxLength) {
                        data.text.delete(maxLength, data.text.length());
                        data.cursor = FastMath.min(data.cursor, maxLength);
                    }
                    FontRenderer font = gui.style().font();
                    FontRenderer.TextInfo textInfo =
                            font.render(FontRenderer.to(),
                                    textFilter.filter(data.text.toString()),
                                    size.floatY(), size.floatX());
                    int maxLengthFont = textInfo.length();
                    if (data.text.length() > maxLengthFont) {
                        data.text = data.text
                                .delete(maxLengthFont, data.text.length());
                        data.cursor = FastMath.min(data.cursor, maxLengthFont);
                    }
                    dirty();
                }
            });
        } else {
            data.selectionStart = -1;
            data.cursor = data.text.length();
            focused = false;
        }
    }

    @Override
    public void renderComponent(GL gl, Shader shader, Vector2 size,
            Vector2 pixelSize, double delta) {
        super.renderComponent(gl, shader, size, pixelSize, delta);
        if (active) {
            if (System.currentTimeMillis() / 600 % 2 == 0) {
                Streams.forEach(vaoCursor, mesh -> {
                    mesh.b.bind(gl);
                    mesh.a.render(gl, shader);
                });
            }
        }
        gl.textures().unbind(gl);
        gl.setBlending(BlendingMode.INVERT);
        Streams.forEach(vaoSelection, mesh -> mesh.a.render(gl, shader));
        gl.setBlending(BlendingMode.NORMAL);
    }
}

File: Engine/src/main/kotlin/org/tobi29/scapes/engine/gui/GuiComponentEditableText.kt
/*
 * Copyright 2012-2018 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.engine.gui

import kotlinx.coroutines.experimental.Job
import kotlinx.coroutines.experimental.launch
import kotlinx.coroutines.experimental.yield
import org.tobi29.math.vector.Vector2d
import org.tobi29.scapes.engine.graphics.*
import org.tobi29.stdex.math.clamp
import kotlin.math.min

class GuiComponentEditableText(
    parent: GuiLayoutData,
    text: String,
    private val maxLength: Int,
    private val active: () -> Boolean,
    private val r: Double = 1.0,
    private val g: Double = 1.0,
    private val b: Double = 1.0,
    private val a: Double = 1.0
) : GuiComponentHeavy(parent) {
    private val data = GuiController.TextFieldData()
    private var vaoCursor: List<Pair<Model, Texture>>? = null
    private var vaoSelection: List<Pair<Model, Texture>>? = null
    private var focused = false
    private var updateJob: Job? = null
    var textFilter: (String) -> String = { it }
        set(value) {
            field = value
            dirty()
        }

    constructor(
        parent: GuiLayoutData,
        text: String,
        active: () -> Boolean,
        r: Double = 1.0,
        g: Double = 1.0,
        b: Double = 1.0,
        a: Double = 1.0
    ) : this(parent, text, Int.MAX_VALUE, active, r, g, b, a)

    init {
        data.text.append(text)
        data.cursor = data.text.length
        dirty()
    }

    val isActive: Boolean get() = active()

    var text: String
        get() = synchronized(data) { data.text.toString() }
        set(value) = synchronized(data) {
            if (data.text.toString() != text) {
                data.text.clear()
                data.text.append(text)
                dirty()
            }
        }

    override fun updateMesh(
        renderer: GuiRenderer,
        size: Vector2d
    ) {
        var text = ""
        var selectionStart = 0
        var selectionEnd = 0
        synchronized(data) {
            text = data.text.toString()
            selectionStart = clamp(data.selectionStart, -1, text.length)
            selectionEnd = clamp(data.selectionEnd, 0, text.length)
        }
        val font = gui.style.font
        font.render(
            FontRenderer.to(renderer, r, g, b, a), textFilter(text),
            size.y, size.x
        )
        val batch = GuiRenderBatch(renderer.pixelSize)
        val cursor = clamp(data.cursor, 0, text.length)
        font.render(
            FontRenderer.to(
                batch, 1.0, 1.0, 1.0, 1.0
            ), text.substring(0, cursor) + '|',
            size.y, size.x, cursor, cursor + 1
        )
        vaoCursor = batch.finish()
        vaoSelection = if (selectionStart >= 0) {
            font.render(
                FontRenderer.to(batch, 0.0, 0.0, true, 1.0, 1.0, 1.0, 1.0),
                text, size.y, size.x, selectionStart, selectionEnd
            )
            batch.finish()
        } else {
            null
        }
    }

    override fun updateVisible() {
        synchronized(this) {
            updateJob?.cancel()
            if (!isVisible) return@synchronized
            updateJob = launch(renderExecutor) {
                while (true) {
                    yield() // Wait for next frame
                    update()
                }
            }
        }
    }

    private fun update() {
        if (isActive) {
            if (!focused) {
                engine.guiController.focusTextField({ isActive }, data, false)
                focused = true
            }
        } else if (focused) {
            data.selectionStart = -1
            data.cursor = data.text.length
            focused = false
        }
        if (data.dirty.getAndSet(false)) {
            size()?.let { size ->
                synchronized(data) {
                    if (data.text.length > maxLength) {
                        data.text.delete(maxLength, data.text.length)
                        data.cursor = min(data.cursor, maxLength)
                    }
                    val font = gui.style.font
                    val textInfo = font.render(
                        FontRenderer.to(),
                        textFilter(data.text.toString()),
                        size.y, size.x
                    )
                    val maxLengthFont = textInfo.length
                    if (data.text.length > maxLengthFont) {
                        data.text.delete(maxLengthFont, data.text.length)
                        data.cursor = min(data.cursor, maxLengthFont)
                    }
                }
            }
            dirty()
        }
    }

    public override fun renderComponent(
        gl: GL,
        shader: Shader,
        size: Vector2d,
        pixelSize: Vector2d,
        delta: Double
    ) {
        super.renderComponent(gl, shader, size, pixelSize, delta)
        if (isActive) {
            if (gl.timestamp / 600000000L % 2L == 0L) {
                vaoCursor?.forEach {
                    it.second.bind(gl)
                    it.first.render(gl, shader)
                }
            }
        }
        gl.textureEmpty.bind(gl)
        gl.setBlending(BlendingMode.INVERT)
        vaoSelection?.forEach { it.first.render(gl, shader) }
        gl.setBlending(BlendingMode.NORMAL)
    }
}
